src/x/lockfile/lockfile.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package lockfile
+
+import (
+	"os"
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+type Lockfile struct {
+	file os.File
+}
+
+// Create creates the given file path if it doesn't exist and
+// obtains an exclusive lock on it. An error is returned if the lock
+// has been obtained by another process.
+func Create(path string) (*Lockfile, error) {
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0666)
```

Comment:
what happens if some of the intermediary required directories don't exist? Will this still work or do you need a call to `os.MkDirAll()`

Reply:
the caller will have to call `MkDirAll()` or he won't be able to obtain the lock

Reply:
Can we add it here so callers don't have to think about it? Maybe can even be an additional function:

```golang
func CreateAndAcquire(path string, newDirMode os.FileMode) (*LockFile, error) {
    if err := os.MkdirAll(paths.Dir(path), newDirMode); err != nil {
        return nil, err
    }

    return Acquire(path)
}
```